jmclient/jmclient/websocketserver.py

```diff
@@ -0,0 +1,72 @@
+import json
+from autobahn.twisted.websocket import WebSocketServerFactory, \
+     WebSocketServerProtocol
+from jmbitcoin import human_readable_transaction
+from jmbase import get_log
+
+jlog = get_log()
+
+class JmwalletdWebSocketServerProtocol(WebSocketServerProtocol):
+    def onOpen(self):
+        self.factory.register(self)
+
+    def sendTxNotification(self, jsontx, txid):
+        """ Pass in a json-formatted transaction details object
+        and a txid (which is technically redundant; the latter
+        is inside the former; but this functions as a unique ID).
+        """
+        if not self.token:
```

Comment:
Yeah probably just  add `self.token = None` in `onOpen` or an overridden protocol instance constructor.

Reply:
So the client must send the authentication message (see `JMwalletdWebScoketServerProtocol.onMessage`) before being able to receive notifications, but yeah clearly this was coded wrong since it should gracefully just not send the notification if you haven't authenticated yet. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/996/commits/b96f4ad40716f0049efe5c12aeb3233536711b09